### PR TITLE
Instrument mongoose_wpool

### PR DESCRIPTION
**Pull request by gustawlippa**
This PR instruments `mongoose_wpool` modules.

Some notes:
- `mongoose_wpool_rdbms` is really tested in the `rdbms_SUITE` when it comes to big tests, and there is a functionality of wrapped requests that was using `mongoose_metrics` in the wrapper. I think this should be changed to a span, but I didn't want to make the scope bigger as this can be done later. Now the tests emits an instrumentation event in the wrapper.
- I decided to remove the metrics test from `mam_SUITE` as this functionality is tested in `rdbms_SUITE` and isn't really about MAM logically - MAM could use a different backend after all.
- I touched docs a bit, as this may make the subsequent work easier, but the way we describe metrics will have to be revisited to reflect Prometheus and Exometer backends (or just the former if we drop Exometer).
- I noticed a slow test and fixed it, as it was a small change.
- in `mongoose_rabbit_worker` the argument to many functions is actually a `HostType` - the host to which Rabbit worker connects is in `amqp_client_opts`, so I changed the name in the whole module.
- in most modules from `mongoose_wpool` I think the argument `HostType` should be renamed to `Scope` - as they can have a very proper and supported value of `global` (for example services use these global pools). I didn't do this now, as it's outside of the scope and didn't want to make the diff huge.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/223253) / elasticsearch_and_cassandra_mnesia / 86974b19e364185ea0cce1a8ff67645bdf6b0d0f
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4306/223253/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4306/223253/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%4000499ef00859.2024-06-18_12.33.35/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/223685) / elasticsearch_and_cassandra_mnesia / 32b335661f778489fbe9abb1d55dd6daa392db3a
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4306/223685/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4306/223685/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%403a51faed053b.2024-06-19_12.50.21/index.html)
**OK: 455** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/223914) / elasticsearch_and_cassandra_mnesia / 170fa7ca5201f0a365bbc5d13856c4839292ff6e
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4306/223914/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4306/223914/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40b7f708d65e07.2024-06-20_10.27.26/index.html)
**OK: 455** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/223948) / elasticsearch_and_cassandra_mnesia / 77c6a95bdbb5a6a55e5571830319de7e2f252b80
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4306/223948/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4306/223948/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40dc708135c3c0.2024-06-20_11.33.22/index.html)
**OK: 455** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/224767) / elasticsearch_and_cassandra_mnesia / fba25c97bd57f6786cb818086c00c9fe75c115e0
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4306/224767/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4306/224767/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%4011c7f39c168e.2024-06-25_12.58.26/index.html)
**OK: 455** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/224804) / elasticsearch_and_cassandra_mnesia / a98cba3f60e3734b9175577a694bcfb24b027d8e
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4306/224804/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4306/224804/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40869d4e5fda7e.2024-06-25_13.15.11/index.html)
**OK: 455** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/224886) / elasticsearch_and_cassandra_mnesia / f1e5803acfd9ff4c630f4bbdc328a997678ad24b
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4306/224886/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4306/224886/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40ff6263c54800.2024-06-25_14.55.02/index.html)
**OK: 455** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/224924) / elasticsearch_and_cassandra_mnesia / 18ead6aad0f14ed82c94c99495174a4b026919fa
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4306/224924/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4306/224924/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40cab206f66e21.2024-06-26_08.21.47/index.html)
**OK: 455** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/225030) / elasticsearch_and_cassandra_mnesia / 77179f57e67806131b6f80a361ed92dd46e59959
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4306/225030/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4306/225030/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%4079eb70800e7c.2024-06-26_10.51.36/index.html)
**OK: 455** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

File: src/wpool/mongoose_wpool_rdbms.erl [hunk 1]
```diff
@@ -1,9 +1,13 @@
...
 
--export([init/0]).
--export([start/4]).
--export([stop/2]).
+-include("mongoose_metrics_definitions.hrl").
+
+-export([init/0, start/4, stop/2]).
+-export([probe/2, instrumentation/2]).
+
+-ignore_xref([instrumentation/2]). % optional wpool callback
```
[the remark this answers is not in the document]
Comment (gustawlippa) [reply src/wpool/mongoose_wpool_rdbms.erl:10]: Hmm, I thought I saw an error here, and that's why I've added it - I was also surprised. But now I can't reproduce it, so it might have been something else. I will remove this line, thanks.

File: src/mongoose_rabbit_worker.erl [hunk 2]
```diff
@@ -197,40 +194,44 @@ maybe_wait_for_confirms(Channel, true) ->
...
 -spec close_rabbit_connection(Connection :: pid(), Channel :: pid(),
-                              Host :: jid:server(), PoolTag :: atom()) ->
+                              HostType :: mongooseim:host_type_or_global(), PoolTag :: atom()) ->
     ok | no_return().
-close_rabbit_connection(Connection, Channel, Host, PoolTag) ->
-    update_closed_connections_metrics(Host, PoolTag),
+close_rabbit_connection(Connection, Channel, HostType, PoolTag) ->
+    mongoose_instrument:execute(?CONNECTIONS, #{host_type => HostType, pool_tag => PoolTag},
+                                #{active => -1, closed => 1}),
+    ?LOG_WARNING(#{what => mongoose_rabbit_worker_terminating_really, ht => HostType, pt => PoolTag}),
```
[the remark this answers is not in the document]
Comment (gustawlippa) [reply src/mongoose_rabbit_worker.erl:234]: Oops, yes, thanks, I'll fix it.

File: src/mongoose_rabbit_worker.erl [hunk 3]
```diff
@@ -37,28 +37,20 @@
...
--define(CONNECTIONS_FAILED_METRIC(Tag), [backends, Tag, connections_failed]).
--define(MESSAGES_PUBLISHED_METRIC(Tag), [backends, Tag, messages_published]).
--define(MESSAGES_FAILED_METRIC(Tag), [backends, Tag, messages_failed]).
--define(MESSAGES_TIMEOUT_METRIC(Tag), [backends, Tag, messages_timeout]).
--define(MESSAGE_PUBLISH_TIME_METRIC(Tag),
-        [backends, Tag, message_publish_time]).
--define(MESSAGE_PAYLOAD_SIZE_METRIC(Tag),
-        [backends, Tag, message_payload_size]).
+-define(CONNECTIONS, mongoose_wpool_rabbit_connections).
+-define(MESSAGES_PUBLISHED, mongoose_wpool_rabbit_messages_published).
```
[the remark this answers is not in the document]
Comment (gustawlippa) [reply src/mongoose_rabbit_worker.erl:53]: I used `mongoose_wpool` as a prefix, because these events are for `mongoose_wpool_rabbit`'s sake and for example for modules wejust used the module names as the event prefix. Would just `wpool_rabbit_` as a prefix be better? I'd think then it'd be better to change `mongoose_wpool_rdbms` events similarly.

Good note about the macros - now that I think about it I agree - I will get rid of them.

File: include/mongoose_metrics_definitions.hrl [hunk 4]
```diff
@@ -57,3 +57,10 @@
                            {send_pend,0}
                           ]).
 
+-define(INET_STATS_METRICS, #{recv_oct => spiral,
```
[the remark this answers is not in the document]
Comment (gustawlippa) [reply include/mongoose_metrics_definitions.hrl:60]: I wanted to put it near `INET_STATS`, as this is what `mongoose_metrics` uses to get the values. I guess I just used the convention that's in the header file. I can change it to a function if it's better.
Or do you think it could go somewhere else in the code entirely?

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/225122) / elasticsearch_and_cassandra_mnesia / e30286d6e4a8545eea83f6d35d70107eb9008ee1
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4306/225122/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4306/225122/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40f1bfabb956a3.2024-06-26_15.30.54/index.html)
**OK: 455** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

File: src/mongoose_rabbit_worker.erl (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/mongoose_rabbit_worker.erl:53]: I'd just skip all `mongoose` prefixes. Of course we could do it later, but having simple prefixes like `auth_`, `sm_`, `router_`, etc. seems more consistent. Also, when browsing metrics ordered alphabetically, you don't get some of them bunched up at `mongoose_`.

File: test/mongoose_rabbit_worker_SUITE.erl [hunk 5]
```diff
@@ -184,12 +184,17 @@ mock_amqp() ->
                 end).
 
 mock_metrics() ->
-    meck:new(mongoose_metrics, [no_link, passthrough]),
-    meck:expect(mongoose_metrics, update, fun(_, _, _) -> ok end).
+    meck:new(mongoose_instrument, [no_link, passthrough]),
```
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply test/mongoose_rabbit_worker_SUITE.erl:187]: I would just start it and not care about `instrument_helper` - after all, the events should be tested anyway in big tests, right? But that's my point of view.

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/225503) / elasticsearch_and_cassandra_mnesia / d7e2ee09e566acbbc9c3c8c319ecdb329f239f33
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4306/225503/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4306/225503/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%406ad49892de65.2024-07-01_09.33.59/index.html)
**OK: 455** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/225538) / elasticsearch_and_cassandra_mnesia / f347ef872d26b95488b8f34831634206a96f9c22
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4306/225538/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4306/225538/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40c646afb79d72.2024-07-01_10.03.52/index.html)
**OK: 455** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/225571) / elasticsearch_and_cassandra_mnesia / be4d19cc188dcecac17ba91cfe8798fe2e6bf783
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4306/225571/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4306/225571/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40ea1b97a73d19.2024-07-01_10.30.09/index.html)
**OK: 455** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

File: src/wpool/mongoose_wpool_rdbms.erl [hunk 6]
```diff
@@ -29,11 +31,30 @@ stop(_, _) ->
...
+    [{wpool_rdbms_stats, #{host_type => HostType, pool_tag => Tag},
+      #{probe => #{module => ?MODULE}, metrics => ?INET_STATS_METRICS#{workers => counter}}}].
+
+-spec probe(mongoose_instrument:event_name(), mongoose_instrument:labels()) ->
+    mongoose_instrument:measurements().
+probe(wpool_global_rdbms_stats, #{pool_tag := Tag} = _Labels) ->
+    Stats = mongoose_metrics:get_rdbms_data_stats([{rdbms, global, Tag}]),
+    proplists:to_map(Stats);
+probe(wpool_rdbms_stats, #{host_type := HostType, pool_tag := Tag} = _Labels) ->
+    Stats = mongoose_metrics:get_rdbms_data_stats([{rdbms, HostType, Tag}]),
```
Comment (chrzaszcz) [inline src/wpool/mongoose_wpool_rdbms.erl:59]: The `mongoose_metrics` module will be removed soon, so we need to move this functionality out of that module. It could be a separate PR, but IMO it's the same story.

File: src/mongoose_rabbit_worker.erl [hunk 7]
```diff
@@ -72,16 +57,13 @@
...
-     {?CONNECTIONS_OPENED_METRIC(Tag), spiral},
-     {?CONNECTIONS_CLOSED_METRIC(Tag), spiral},
-     {?CONNECTIONS_FAILED_METRIC(Tag), spiral},
-     {?MESSAGES_PUBLISHED_METRIC(Tag), spiral},
-     {?MESSAGES_FAILED_METRIC(Tag), spiral},
-     {?MESSAGES_TIMEOUT_METRIC(Tag), spiral},
-     {?MESSAGE_PUBLISH_TIME_METRIC(Tag), histogram},
-     {?MESSAGE_PAYLOAD_SIZE_METRIC(Tag), histogram}].
+-spec list_metrics(mongooseim:host_type_or_global(), mongoose_wpool:tag()) -> [mongoose_instrument:spec()].
+list_metrics(HostType, Tag) ->
```
Comment (chrzaszcz) [inline src/mongoose_rabbit_worker.erl:61]: We could call it `instrumentation` for consistency.

**Review by chrzaszcz — COMMENTED** [review]
Looks good. I have a few minor comments.

File: big_tests/tests/mod_event_pusher_rabbit_SUITE.erl [hunk 8]
```diff
@@ -694,3 +700,14 @@ is_rabbitmq_available() ->
         _Err ->
             false
     end.
+
+assert_connection_events() ->
+    instrument_helper:assert(wpool_rabbit_connections,
+                             #{host_type => domain(), pool_tag => event_pusher},
+                             fun(#{active := 1, opened := 1}) -> true end),
+    Measurements = instrument_helper:wait_for(wpool_rabbit_connections,
```
[the remark this answers is not in the document]
Comment (gustawlippa) [reply big_tests/tests/mod_event_pusher_rabbit_SUITE.erl:708]: Ah! That's a good point indeed. It's tricky not to introduce a race condition then. I think I will have to split the connect/disconnect assertions, and stop the rabbit wpool as well as and use `instrument_helper:take/2`between them.

File: src/wpool/mongoose_wpool_rdbms.erl (hunk 6 above)
Comment (gustawlippa) [reply src/wpool/mongoose_wpool_rdbms.erl:59]: I was on the fence with this, as I wasn't sure where will `mongoose_metrics:inet_stats/1` live after we remove `mongoose_metrics`. But I can move this out, and change the macro from the earlier comment to a function like you've proposed, as well as move it out of the `mongoose_metrics.hrl` 👍

File: big_tests/tests/mod_event_pusher_rabbit_SUITE.erl (hunk 8 above)
[the remark this answers is not in the document]
Comment (gustawlippa) [reply big_tests/tests/mod_event_pusher_rabbit_SUITE.erl:714]: I think I changed it to a better version now, without the problem that `wait_for` wasn't really waiting for anything.

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/225679) / elasticsearch_and_cassandra_mnesia / b7145f8ebd0cadc368615cf1632b38b34948a38e
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4306/225679/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4306/225679/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40cd76072f807b.2024-07-02_08.21.23/index.html)
**OK: 455** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/225701) / elasticsearch_and_cassandra_mnesia / 5c27ad5a0bbb2cb1e5d4f318a5688e8e0613b3f9
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4306/225701/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4306/225701/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40946d22dfe651.2024-07-02_08.35.30/index.html)
**OK: 455** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/225814) / elasticsearch_and_cassandra_mnesia / 5fea3e739346e734204d56ed11b393cfabdf3234
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4306/225814/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4306/225814/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%405d02514e1664.2024-07-02_12.27.49/index.html)
**OK: 455** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Review by chrzaszcz — APPROVED** [review]
Looks good 👍
